presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java

```diff
@@ -88,10 +90,28 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
                 jdbcHandle.getCatalogName(),
                 jdbcHandle.getSchemaName(),
                 jdbcHandle.getTableName(),
+                jdbcHandle.getLimit(),
                 constraint.getSummary());
         return Optional.of(new ConstraintApplicationResult<>(jdbcHandle, constraint.getSummary()));
     }
 
+    @Override
+    public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(ConnectorTableHandle handle, long limit)
+    {
+        JdbcTableHandle jdbcHandle = (JdbcTableHandle) handle;
+        if (jdbcHandle.getLimit().isPresent() && jdbcHandle.getLimit().getAsLong() <= limit) {
+            return Optional.empty();
+        }
+        jdbcHandle = new JdbcTableHandle(
+                jdbcHandle.getSchemaTableName(),
+                jdbcHandle.getCatalogName(),
+                jdbcHandle.getSchemaName(),
+                jdbcHandle.getTableName(),
+                OptionalLong.of(limit),
+                jdbcHandle.getTupleDomain());
+        return Optional.of(new LimitApplicationResult<>(jdbcHandle, false));
```

Comment:
Why is `limitGuaranteed` false? Is that because connectors might not implement ``BaseJdbcClient::applyLimit``, so we can't be sure? In that case, add a comment, as the relationship between this line and that method is non-obvious.

It'd be better to come up with a way for this to know whether connector implementations handle the limit properly and produce a precise response for `limitGuaranteed`. Otherwise, the engine will have to preserve a global limit above the table scan, which will prevent it from pushing down any other operations that might be sitting on top of the limit.

Reply:
This is implemented now by adding a `isLimitGuaranteed()` method to `JdbcClient`. I don't like how this method is separate from the application of it, but I don't see a better way without rethinking the whole API.

Reply:
> I don't like how this method is separate from the application of it, but I don't see a better way without rethinking the whole API.

Consider replacing `io.prestosql.plugin.jdbc.BaseJdbcClient#limitFunction` with:
```
protected Optional<BiFunction<String, Long, String>> limitFunction()
{
    return Optional.empty();
}
```

Then you know whether subclass supports limiting or not and can properly fill `isLimitGuaranteed`.

If limiting support might depend on limit value (for example 32-bit int vs long), we can go with:
```
protected Optional<Function<String, String>> limitFunction(long limit)
{
    return Optional.empty();
}
```